net_io.c

```diff
@@ -776,13 +782,270 @@ static void send_sbs_heartbeat(struct net_service *service)
+    if (!p)
+        return;
+
+    // NOTE: not sure about message types, so keep the original code from
+    // stratux/dump1090 as-is.
+    //
+    // Decide on the basic SBS Message Type
+    if        ((mm->msgtype ==  4) || (mm->msgtype == 20)) {
+        msgType = 5;
+    } else if ((mm->msgtype ==  5) || (mm->msgtype == 21)) {
+        msgType = 6;
+    } else if ((mm->msgtype ==  0) || (mm->msgtype == 16)) {
+        msgType = 7;
+    } else if  (mm->msgtype == 11) {
+        msgType = 8;
+    } else if ((mm->msgtype != 17) && (mm->msgtype != 18)) {
+        return;
+    } else if ((mm->metype >= 1) && (mm->metype <=  4)) {
+        msgType = 1;
+    } else if ((mm->metype >= 5) && (mm->metype <=  8)) {
+        if (mm->cpr_decoded)
+            {msgType = 2;}
+        else
+            {msgType = 7;}
+    } else if ((mm->metype >= 9) && (mm->metype <= 18)) {
+        if (mm->cpr_decoded)
+            {msgType = 3;}
+        else
+            {msgType = 7;}
+    } else if ((mm->metype == 29) || (mm->metype == 31)) {
+        msgType = 9; // new message type for target state and operational status messages
+    } else if (mm->metype !=  19) {
+        return;
+    } else if ((mm->mesub == 1) || (mm->mesub == 2)) {
+        msgType = 4;
+    } else {
+        return;
+    }
+
+    // Begin populating the traffic.go fields.
+    // ICAO address, Mode S message types, and signal level
+
+    int cacf = 0; // overload the JSON "CA" field to report CA (DF11 or DF17), CF (DF18), or zero (all other DF types)
+    if ((mm->msgtype == 11) || (mm->msgtype == 17)) {
+        cacf = mm->CA;
+    } else if (mm->msgtype == 18) {
+        cacf = mm->CF;
+    }
+
+    p += sprintf(p,
+            "{\"Icao_addr\":%d,"
+            "\"DF\":%d,\"CA\":%d,"
+            "\"TypeCode\":%d,"
+            "\"SubtypeCode\":%d,"
+            "\"SBS_MsgType\":%d,"
+            "\"SignalLevel\":%f,",
+            mm->addr,
+            mm->msgtype, cacf,
+            mm->metype,
+            mm->mesub,
+            msgType,
+            mm->signalLevel); // what precision and range is needed for RSSI?
+
+    // Find current system time
+    clock_gettime(CLOCK_REALTIME, &now);
+    gmtime_r(&now.tv_sec, &stTime_now);  // we expect UTC
+
+    // Find message reception time
+    time_t received = (time_t) (mm->sysTimestampMsg / 1000);
+    localtime_r(&received, &stTime_receive);
+
+    //// callsign
+    if (mm->callsign_valid)
+        p += sprintf(p, "\"Tail\":\"%s\",", mm->callsign);
+    else
+        p += sprintf(p, "\"Tail\":null,");
+
+    //// altitude & gnss
+    bool alt_is_geom = false;
+    if (Modes.use_gnss && mm->altitude_geom_valid) {
+        p += sprintf(p, "\"Alt\":%d,",mm->altitude_geom);
+        alt_is_geom = true;
+    } else if (mm->altitude_baro_valid)
+        p += sprintf(p, "\"Alt\":%d,",mm->altitude_baro);
+    else
+        p += sprintf(p, "\"Alt\":null,");
```

Comment:
This probably needs a little rework. Given that stratux is being told whether the given altitude is barometric or geometric already (see the next bit of code) I'd guess that the thing to do here is:

1) if baro valid, emit baro alt;
2) else if geom valid, emit geom alt;
3) emit AltIsGNSS reflecting what sort of thing was emitted;
4) don't be conditional on Modes.use_gnss at all.